### PR TITLE
Use inline static member variables for CudaInternal

### PR DESCRIPTION
**Pull request by masterleinad**
Addressing https://github.com/kokkos/kokkos/pull/5442#discussion_r974176530 in correspondence of doing the same changes to the `HIP` backend in [`891fe55` (#5444)](https://github.com/kokkos/kokkos/pull/5444/commits/891fe55638db9f4c59d661b24728af5a1a6fb331).

### CODE REVIEW

**Review by dalg24 — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp [hunk 1]
```diff
@@ -129,9 +129,9 @@ class CudaInternal {
 
   // FIXME_CUDA: these want to be per-device, not per-stream...  use of 'static'
   //  here will break once there are multiple devices though
-  static unsigned long* constantMemHostStaging;
-  static cudaEvent_t constantMemReusable;
-  static std::mutex constantMemMutex;
+  inline static unsigned long* constantMemHostStaging = nullptr;
+  inline static cudaEvent_t constantMemReusable       = nullptr;
+  inline static std::mutex constantMemMutex;
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Instance.hpp:134]: Comment L130-131 looks outdated

**Review by cz4rs — APPROVED** [review]

**Comment by dalg24** [comment]
What did you change?

**Comment by masterleinad** [comment]
> What did you change?

I fixed a typo. `cudaDeviceProp_t` -> `cudaDeviceProp`.

**Comment by masterleinad** [comment]
All CUDA CI passes.

**Comment by jczhang07** [comment]
@masterleinad @dalg24 

I met an error in petsc/kokkos after this PR.  I created a standalone example [visibility.tgz](https://github.com/kokkos/kokkos/files/9913696/visibility.tgz).  I tested with gcc, icc, and clang. They all showed the same problem.

A "hello world" library
$ cat hw.hpp 
```
class MyClass {
public:
  // __attribute__ ((visibility ("default"))) 
  static inline int member = 1;
};

extern "C" void foo();
```
$ cat hw.cpp
```
#include <stdio.h>
#include <hw.hpp>

void foo() {
  MyClass obj;

  obj.member = 100;
  printf("In libhw, obj.member (%p) = %d\n", &obj.member, obj.member);
}
```
$ cat test.cpp
```
#include <stdio.h>
#include <hw.hpp>

int main() {
  MyClass a;

  printf("In main, a.member (%p) = %d\n", &a.member, a.member);
  foo();
  printf("In main, MyClass::member (%p) = %d\n", &MyClass::member, MyClass::member);
  return 0;
}
```



$ cat makefile 
```
CC=gcc
#CC=clang++
#CC=icpc

VISFLAG= -fvisibility=hidden
#VISFLAG=

CXXFLAGS+=-std=c++17

DIR:=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))

all : clean run

hw.o : hw.cpp
        $(CC) $(CXXFLAGS) -c -fPIC -I$(DIR) hw.cpp
libhw.so : hw.o
        $(CC) -o libhw.so -shared hw.o

test.o : test.cpp
        $(CC) $(CXXFLAGS) -c -fPIC -I$(DIR) test.cpp $(VISFLAG)
test : test.o libhw.so
        $(CC) $(CXXFLAGS) -o test test.o -Wl,-rpath $(DIR) -L$(DIR) -lhw

clean:
        rm -f *.o *.so test
run : test
        ./test

.PHONY: all run clean
```
With the `VISFLAG`
```
$ make
rm -f *.o *.so test
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility test.cpp 
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility hw.cpp
gcc -o libhw.so -shared hw.o
gcc -std=c++17 -o test test.o -Wl,-rpath /home/jczhang/test/visibility -L/home/jczhang/test/visibility -lhw
./test
In main, a.member (0x5654704a8010) = 1
In libhw, obj.member (0x5654704a8010) = 100
In main, MyClass::member (0x5654704a8010) = 100
```

Comment out the `VISFLAG`

```
$ make
rm -f *.o *.so test
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility test.cpp -fvisibility=hidden
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility hw.cpp
gcc -o libhw.so -shared hw.o
gcc -std=c++17 -o test test.o -Wl,-rpath /home/jczhang/test/visibility -L/home/jczhang/test/visibility -lhw
./test
In main, a.member (0x560eacbf8010) = 1
In libhw, obj.member (0x7f440e6b2028) = 100
In main, MyClass::member (0x560eacbf8010) = 1
```

Adding `__attribute__ ((visibility ("default"))) ` in hw.cpp will fix the problem.  But I don't really understand the issue.  Does `-fvisibility=hidden` override external linkage of the static variable?

**Comment by jczhang07** [comment]
@balay  Separate from the C++ 17 new feature `static line variable`, if I compile hw.cpp also with `-fvisibility=hidden`, I will met the link error

```
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility test.cpp -fvisibility=hidden
gcc -std=c++17 -c -fPIC -I/home/jczhang/test/visibility hw.cpp -fvisibility=hidden
gcc -o libhw.so -shared hw.o
gcc -std=c++17 -o test test.o -Wl,-rpath /home/jczhang/test/visibility -L/home/jczhang/test/visibility -lhw
/usr/bin/ld: test.o: in function `main':
test.cpp:(.text+0x33): undefined reference to `foo'
```
I am confused since from `man gcc`,

> "extern" declarations are not affected by -fvisibility, so a lot of code can be recompiled with -fvisibility=hidden with no modifications.

**Comment by jczhang07** [comment]
@dalg24 I filed a gcc bug at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107535.
Our workaround in petsc is to compile petsc source files including kokkos headers without `-fvisibility=hidden`, but other files are still compiled with the flag. I can imagine other codes using kokkos and compiling with `-fvisibility=hidden` will also run into the same problem.

**Pull request merged**